// ng2-ya-table/dist/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var forms_1 = require('@angular/forms');
var pagination_1 = require('ng2-bootstrap/pagination');
var ng2_ya_table_directives_1 = require('./src/ng2-ya-table-directives');
var ng2_ya_table_service_1 = require('./src/ng2-ya-table/ng2-ya-table.service');
var Ng2YaTableModule = (function () {
    function Ng2YaTableModule() {
    }
    Ng2YaTableModule.forRoot = function () {
        return {
            ngModule: Ng2YaTableModule,
            providers: [ng2_ya_table_service_1.Ng2YaTableService]
        };
    };
    Ng2YaTableModule = __decorate([
        core_1.NgModule({
            imports: [
                common_1.CommonModule,
                forms_1.FormsModule,
                pagination_1.PaginationModule.forRoot(),
            ],
            declarations: [
                ng2_ya_table_directives_1.NG2_POWER_TABLE_DIRECTIVES
            ],
            exports: [
                ng2_ya_table_directives_1.NG2_POWER_TABLE_DIRECTIVES
            ]
        }), 
        __metadata('design:paramtypes', [])
    ], Ng2YaTableModule);
    return Ng2YaTableModule;
}());
exports.Ng2YaTableModule = Ng2YaTableModule;
