// ng2-ya-table/dist/src/ng2-ya-table/ng2-ya-table-filtering-default.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var Subject_1 = require('rxjs/Subject');
require("rxjs/Rx");
var ng2_ya_table_service_1 = require('./ng2-ya-table.service');
var Ng2YaTableFilteringDefaultComponent = (function () {
    function Ng2YaTableFilteringDefaultComponent(state) {
        var _this = this;
        this.state = state;
        this.filterValueChanged = new Subject_1.Subject();
        this.filterValueChanged
            .debounceTime(300)
            .distinctUntilChanged()
            .subscribe(function (filterValue) { return _this.state.changeFilter(_this.column); });
    }
    Ng2YaTableFilteringDefaultComponent.prototype.onFilterValueChange = function (event) {
        this.filterValueChanged.next(event);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], Ng2YaTableFilteringDefaultComponent.prototype, "column", void 0);
    Ng2YaTableFilteringDefaultComponent = __decorate([
        core_1.Component({
            selector: 'ng2-ya-table-filter-default',
            template: "\n    <input class=\"form-control\" style=\"width: 100%\"\n      [attr.type]=\"column.def.filter.config && column.def.filter.config.type ? column.def.filter.config.type : 'text'\"\n      [attr.placeholder]=\"column.def.filter.config && column.def.filter.config.placeholder ? column.def.filter.config.placeholder : ''\"\n      [attr.max]=\"column.def.filter.config && column.def.filter.config.max ? column.def.filter.config.max : 524288\"\n      [attr.min]=\"column.def.filter.config && column.def.filter.config.min ? column.def.filter.config.min : 0\"\n      [attr.step]=\"column.def.filter.config && column.def.filter.config.step ? column.def.filter.config.step : 1\"\n      [(ngModel)]=\"column.filterValue\"\n      (ngModelChange)=\"onFilterValueChange($event)\" />"
        }), 
        __metadata('design:paramtypes', [ng2_ya_table_service_1.Ng2YaTableService])
    ], Ng2YaTableFilteringDefaultComponent);
    return Ng2YaTableFilteringDefaultComponent;
}());
exports.Ng2YaTableFilteringDefaultComponent = Ng2YaTableFilteringDefaultComponent;
