// ng2-ya-table/dist/src/ng2-ya-table/ng2-ya-table.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var platform_browser_1 = require('@angular/platform-browser');
var Subject_1 = require('rxjs/Subject');
require('rxjs/add/operator/debounceTime');
var ng2_ya_table_service_1 = require('./ng2-ya-table.service');
var Ng2YaTableComponent = (function () {
    function Ng2YaTableComponent(sanitizer, state) {
        var _this = this;
        this.sanitizer = sanitizer;
        this.state = state;
        this.fullTextFilterValueChanged = new Subject_1.Subject();
        this.processing = false;
        this.options = null;
        this.rows = [];
        this.datasource = null;
        this.columns = [];
        this.paging = null;
        this.fullTextFilterValueChanged
            .debounceTime(300)
            .distinctUntilChanged()
            .subscribe(function (filterValue) { return _this.onChangeTable(); });
    }
    Ng2YaTableComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.state.stateChanged$.subscribe(function () { return _this.onChangeTable(); });
    };
    Ng2YaTableComponent.prototype.ngOnChanges = function (changes) {
        if (changes['options'].isFirstChange()) {
            this.state.setOptions(changes['options'].currentValue);
        }
        if (changes['paging'].isFirstChange()) {
            this.state.setPaging(changes['paging'].currentValue);
        }
        if (changes['columns'].isFirstChange()) {
            this.state.setColumns(changes['columns'].currentValue);
        }
    };
    Ng2YaTableComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    Ng2YaTableComponent.prototype.sanitize = function (html) {
        return this.sanitizer.bypassSecurityTrustHtml(html);
    };
    Ng2YaTableComponent.prototype.onChangeTable = function () {
        var _this = this;
        if (this.datasource) {
            this.processing = true;
            var orders_1 = new Array();
            this.state.sortStack.forEach(function (column) {
                var order = {
                    dir: column.sortOrder,
                    name: column.def.name
                };
                orders_1.push(order);
            });
            var filters_1 = new Array();
            this.state.columns.forEach(function (column) {
                if (column.hasFilter) {
                    var filter = {
                        name: column.def.name,
                        type: column.def.filter.type,
                        value: column.filterValue
                    };
                    filters_1.push(filter);
                }
            });
            var request = {
                start: (this.state.paging.currentPage - 1) * this.state.paging.itemsPerPage,
                length: this.state.paging.itemsPerPage,
                filters: filters_1,
                orders: orders_1,
                fullTextFilter: this.state.fullTextFilter
            };
            this.datasource(request).subscribe(function (result) {
                _this.rows = result.data;
                _this.state.paging.recordsFiltered = result.recordsFiltered;
                _this.state.paging.recordsTotal = result.recordsTotal;
            }, function (error) {
                console.log(error);
            }, function () {
                _this.processing = false;
            });
        }
    };
    Ng2YaTableComponent.prototype.getHtml = function (row, column) {
        if (column.render) {
            var data = this.getData(row, column.name);
            return column.render(data, row);
        }
        return this.getData(row, column.name);
    };
    Ng2YaTableComponent.prototype.getData = function (row, propertyName) {
        return propertyName.split('.').reduce(function (prev, curr) { return prev[curr]; }, row);
    };
    Ng2YaTableComponent.prototype.cellClick = function (row, column) {
        if (column.action) {
            var data = this.getData(row, column.name);
            column.action(data, row);
        }
    };
    Ng2YaTableComponent.prototype.onFullTextFilterValueChange = function (event) {
        this.fullTextFilterValueChanged.next(event);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], Ng2YaTableComponent.prototype, "options", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], Ng2YaTableComponent.prototype, "rows", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Function)
    ], Ng2YaTableComponent.prototype, "datasource", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Array)
    ], Ng2YaTableComponent.prototype, "columns", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], Ng2YaTableComponent.prototype, "paging", void 0);
    Ng2YaTableComponent = __decorate([
        core_1.Component({
            selector: 'ng2-ya-table',
            template: "\n    <div class=\"ng2-ya-table_wrapper form-inline dt-bootstrap no-footer\">\n      <div class=\"row\">\n        <div class=\"col-md-6\">\n          <div class=\"ng2-ya-table_length\">\n            <label>\n              <span *ngFor=\"let s of state.language.lengthMenu.split(' ')\">\n                <span [ngSwitch]=\"s\">\n                  <select *ngSwitchCase=\"'_MENU_'\" class=\"form-control input-sm\" [(ngModel)]=\"state.paging.itemsPerPage\" (change)=\"state.changePaging(1, $event.target.value)\">\n                    <option *ngFor=\"let pn of paging.itemsPerPageOptions\" [value]=\"pn\">{{pn}}</option>\n                  </select>\n                  <span *ngSwitchDefault>{{s}}</span>\n                </span>\n              </span>\n            </label>\n          </div>\n        </div>\n        <div class=\"col-md-6\">\n          <div class=\"ng2-ya-table_filter\">\n            <label>\n              <span>{{state.language.search}}</span>\n              <input type=\"search\" class=\"form-control input-sm\"\n                [(ngModel)]=\"state.fullTextFilter\"\n                (ngModelChange)=\"onFullTextFilterValueChange($event)\"/>\n            </label>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"row\">\n        <div class=\"col-md-12\">\n          <div *ngIf=\"processing\" class=\"ng2-ya-table_processing panel panel-default\">{{state.language.processing}}</div>\n          <table class=\"table ng2-ya-table no-footer\" ngClass=\"{{options.className || ''}}\" role=\"grid\" style=\"width: 100%;\">\n            <thead>\n              <tr role=\"row\">\n                <th *ngFor=\"let column of state.columns\" \n                  [style.width]=\"column.def.width\"\n                  [ngClass]=\"{'sorting_desc': column.sortOrder === 'desc', 'sorting_asc': column.sortOrder === 'asc', 'sorting': column.hasSort }\"\n                  [ngTableSorting]=\"column\">\n                  {{column.def.title}}\n                </th>\n              </tr>\n            </thead>\n            <tbody>\n              <tr *ngIf=\"state.showFilterRow\">\n                <td *ngFor=\"let column of state.columns\">\n                  <div *ngIf=\"column.def.filter\" [ngSwitch]=\"column.def.filter.controlType\">\n                    <ng2-ya-table-filter-list *ngSwitchCase=\"'list'\" [column]=\"column\"></ng2-ya-table-filter-list>\n                    <ng2-ya-table-filter-default *ngSwitchDefault [column]=\"column\"></ng2-ya-table-filter-default>\n                  </div>\n                </td>\n              </tr>\n              <tr *ngFor=\"let row of rows\">\n                <td (click)=\"cellClick(row, column)\" *ngFor=\"let column of columns\" [innerHtml]=\"sanitize(getHtml(row, column))\"></td>\n              </tr>\n            </tbody>\n          </table>\n        </div>\n      </div>\n\n      <div class=\"row\">\n        <div class=\"col-md-6\">\n          <div *ngIf=\"rows.length > 0\" class=\"ng2-ya-table_info\" role=\"status\">\n            <span *ngFor=\"let s of state.language.info.split(' ')\">\n              <span [ngSwitch]=\"s\">\n                <span *ngSwitchCase=\"'_START_'\">{{(state.paging.currentPage - 1) * state.paging.itemsPerPage + 1}}</span>\n                <span *ngSwitchCase=\"'_END_'\">{{(state.paging.currentPage - 1) * state.paging.itemsPerPage + rows.length}}</span>\n                <span *ngSwitchCase=\"'_TOTAL_'\">{{state.paging.recordsFiltered}}</span>\n                <span *ngSwitchDefault>{{s}}</span>\n              </span>\n            </span>\n          </div>\n        </div>\n        <div class=\"col-md-6\">\n          <div class=\"ng2-ya-table_paginate paging_simple_numbers\">\n            <pagination *ngIf=\"rows.length > 0\"\n              [(ngModel)]=\"state.paging.currentPage\"\n              [totalItems]=\"state.paging.recordsFiltered\"\n              [itemsPerPage]=\"state.paging.itemsPerPage\"\n              [maxSize]=\"paging.maxSize\"\n              [boundaryLinks]=\"false\"\n              [rotate]=\"false\"\n              (pageChanged)=\"state.changePaging($event.page, $event.itemsPerPage)\"\n              [firstText] = \"state.language.pagination.first\"\n              [lastText] = \"state.language.pagination.last\"\n              [nextText] = \"state.language.pagination.next\"\n              [previousText] = \"state.language.pagination.previous\">\n            </pagination>\n          </div>\n        </div>\n      </div>\n    </div>",
            styles: [
                "table.ng2-ya-table {\n      clear: both;\n      margin-top: 6px !important;\n      margin-bottom: 6px !important;\n      max-width: none !important;\n      border-collapse: separate !important;\n    }",
                "table.ng2-ya-table td,\n    table.ng2-ya-table th {\n      -webkit-box-sizing: content-box;\n      box-sizing: content-box;\n    }",
                "table.ng2-ya-table td.ng2-ya-table_empty,\n    table.ng2-ya-table th.ng2-ya-table_empty {\n      text-align: center;\n    }",
                "table.ng2-ya-table.nowrap th,\n    table.ng2-ya-table.nowrap td {\n      white-space: nowrap;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_length label {\n      font-weight: normal;\n      text-align: left;\n      white-space: nowrap;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_length select {\n      width: 75px;\n      display: inline-block;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_filter {\n      text-align: right;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_filter label {\n      font-weight: normal;\n      white-space: nowrap;\n      text-align: left;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_filter input {\n      margin-left: 0.5em;\n      display: inline-block;\n      width: auto;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_info {\n      padding-top: 8px;\n      white-space: nowrap;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_paginate {\n      margin: 0;\n      white-space: nowrap;\n      text-align: right;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_paginate ul.pagination {\n      margin: 2px 0;\n      white-space: nowrap;\n    }",
                "div.ng2-ya-table_wrapper div.ng2-ya-table_processing {\n      position: absolute;\n      top: 50%;\n      left: 50%;\n      width: 200px;\n      margin-left: -100px;\n      margin-top: -26px;\n      text-align: center;\n      padding: 1em 0;\n    }",
                "table.ng2-ya-table thead > tr > th.sorting_asc, \n    table.ng2-ya-table thead > tr > th.sorting_desc, \n    table.ng2-ya-table thead > tr > th.sorting,\n    table.ng2-ya-table thead > tr > td.sorting_asc,\n    table.ng2-ya-table thead > tr > td.sorting_desc,\n    table.ng2-ya-table thead > tr > td.sorting {\n      padding-right: 30px;\n    }",
                "table.ng2-ya-table thead > tr > th:active,\n    table.ng2-ya-table thead > tr > td:active {\n      outline: none;\n    }",
                "table.ng2-ya-table thead .sorting,\n    table.ng2-ya-table thead .sorting_asc,\n    table.ng2-ya-table thead .sorting_desc,\n    table.ng2-ya-table thead .sorting_asc_disabled,\n    table.ng2-ya-table thead .sorting_desc_disabled {\n      cursor: pointer;\n      position: relative;\n    }",
                "table.ng2-ya-table thead .sorting:after,\n    table.ng2-ya-table thead .sorting_asc:after,\n    table.ng2-ya-table thead .sorting_desc:after,\n    table.ng2-ya-table thead .sorting_asc_disabled:after,\n    table.ng2-ya-table thead .sorting_desc_disabled:after {\n      position: absolute;\n      bottom: 8px;\n      right: 8px;\n      display: block;\n      font-family: 'Glyphicons Halflings';\n      opacity: 0.5;\n    }",
                "table.ng2-ya-table thead .sorting:after {\n      opacity: 0.2;\n      content: \"\\e150\";\n      /* sort */\n    }",
                "table.ng2-ya-table thead .sorting_asc:after {\n      content: \"\\e155\";\n      /* sort-by-attributes */\n    }",
                "table.ng2-ya-table thead .sorting_desc:after {\n      content: \"\\e156\";\n      /* sort-by-attributes-alt */\n    }",
                "table.ng2-ya-table thead .sorting_asc_disabled:after,\n    table.ng2-ya-table thead .sorting_desc_disabled:after {\n      color: #eee;\n    }",
                "@media screen and (max-width: 767px) {\n      div.ng2-ya-table_wrapper div.ng2-ya-table_length,\n      div.ng2-ya-table_wrapper div.ng2-ya-table_filter,\n      div.ng2-ya-table_wrapper div.ng2-ya-table_info,\n      div.ng2-ya-table_wrapper div.ng2-ya-table_paginate {\n        text-align: center;\n      }\n    }"],
            providers: [ng2_ya_table_service_1.Ng2YaTableService]
        }), 
        __metadata('design:paramtypes', [platform_browser_1.DomSanitizer, ng2_ya_table_service_1.Ng2YaTableService])
    ], Ng2YaTableComponent);
    return Ng2YaTableComponent;
}());
exports.Ng2YaTableComponent = Ng2YaTableComponent;
